src/plugins/vdo.c

```diff
@@ -864,3 +864,141 @@ gboolean bd_vdo_grow_physical (const gchar *name, const BDExtraArg **extra, GErr
+    g_hash_table_replace (stats, g_strdup ("savings"), g_strdup_printf ("%ld", savings));
+    if (savings >= 0)
+        g_hash_table_replace (stats, g_strdup ("savingPercent"), g_strdup_printf ("%ld", savings));
+}
+
+static void add_journal_stats (GHashTable *stats) {
+    gint64 journal_entries_committed, journal_entries_started, journal_entries_written;
+    gint64 journal_blocks_committed, journal_blocks_started, journal_blocks_written;
+
+    if (! get_stat_val (stats, "journal_entries_committed", &journal_entries_committed) ||
+        ! get_stat_val (stats, "journal_entries_started", &journal_entries_started) ||
+        ! get_stat_val (stats, "journal_entries_written", &journal_entries_written) ||
+        ! get_stat_val (stats, "journal_blocks_committed", &journal_blocks_committed) ||
+        ! get_stat_val (stats, "journal_blocks_started", &journal_blocks_started) ||
+        ! get_stat_val (stats, "journal_blocks_written", &journal_blocks_written))
+        return;
+
+    g_hash_table_replace (stats, g_strdup ("journal_entries_batching"), g_strdup_printf ("%ld", journal_entries_started - journal_entries_written));
+    g_hash_table_replace (stats, g_strdup ("journal_entries_writing"), g_strdup_printf ("%ld", journal_entries_written - journal_entries_committed));
+    g_hash_table_replace (stats, g_strdup ("journal_blocks_batching"), g_strdup_printf ("%ld", journal_blocks_started - journal_blocks_written));
+    g_hash_table_replace (stats, g_strdup ("journal_blocks_writing"), g_strdup_printf ("%ld", journal_blocks_written - journal_blocks_committed));
+}
+
+static void add_computed_stats (GHashTable *stats) {
+    const gchar *s;
+
+    s = g_hash_table_lookup (stats, "logical_block_size");
+    g_hash_table_replace (stats,
+                          g_strdup ("fiveTwelveByteEmulation"),
+                          g_strdup ((g_strcmp0 (s, "512") == 0) ? "true" : "false"));
+
+    add_write_ampl_r_stats (stats);
+    add_block_stats (stats);
+    add_journal_stats (stats);
+}
+
+/**
+ * bd_vdo_get_statistics:
+ * @name: name of an existing VDO volume
+ * @extra: (allow-none) (array zero-terminated=1): extra options, currently unused
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: (transfer full) (element-type utf8 utf8): hashtable of type string - string of available statistics
+ *
+ * Statistics are collected from the values exposed by the kernel `kvdo` module at the `/sys/kvdo/<VDO_NAME>/statistics/` path. Some of the keys are computed to mimic the information produced by the vdo tools.
+ *
+ * Tech category: %BD_VDO_TECH_VDO-%BD_VDO_TECH_MODE_QUERY
+ */
+GHashTable* bd_vdo_get_statistics (const gchar *name, const BDExtraArg **extra G_GNUC_UNUSED, GError **error) {
+    GHashTable *stats;
+    GDir *dir;
+    gchar *stats_dir;
+    const gchar *direntry;
+    gchar *s;
+    gchar *val;
+
+    if (!check_deps (&avail_deps, DEPS_VDO_MASK, deps, DEPS_LAST, &deps_check_lock, error) ||
```

Comment:
The statistics are provided by the kernel module, so this function should work without the vdo cli tools, right?